controllers/metal3remediation_controller.go

```diff
@@ -222,6 +220,250 @@ func (r *Metal3RemediationReconciler) reconcileDelete(ctx context.Context,
 	return ctrl.Result{}, nil
 }
 
+// remediateRebootStrategy executes the remediation using the reboot strategy.
+// Returns nil, nil when reconcile can continue.
+// Return a Result and optionally an error when reconcile should return.
+func (r *Metal3RemediationReconciler) remediateRebootStrategy(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface) (ctrl.Result, error) {
+	// try to get node
+	node, err := remediationMgr.GetNode(ctx)
+	if err != nil && !apierrors.IsNotFound(err) {
+		r.Log.Error(err, "error getting node for remediation")
+		return ctrl.Result{}, errors.Wrap(err, "error getting node for remediation")
+	}
+
+	switch remediationMgr.GetRemediationPhase() {
+	case capm3.PhaseStarting:
+
+		// add finalizer
+		if set := remediationMgr.SetFinalizer(); set {
+			now := metav1.Now()
+			remediationMgr.SetLastRemediationTime(&now)
+			remediationMgr.IncreaseRetryCount()
```

Comment:
The finalizer (which is in the object metadata) and Status data (in a subresource) are written to the API by [two separate API calls](https://github.com/kubernetes-sigs/cluster-api/blob/main/util/patch/patch.go#L142-L143).
Therefore it's not safe to rely on the state from the first one to ensure the second one only happens once. If the second write fails then it will never be repeated.

I think you want to update the count/time atomically with the change into the Starting phase.

Reply:
good catch, thanks!